util/network_linux.go

```diff
@@ -0,0 +1,136 @@
+// +build linux
+
+package util
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"context"
+	"net"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/retry"
+	"github.com/DataDog/tcptracer-bpf/pkg/tracer"
+)
+
+const (
+	statusURL      = "http://unix/status"
+	connectionsURL = "http://unix/connections"
+)
+
+var (
+	globalUtil            *RemoteNetTracerUtil
+	globalSocketPath      string
+	hasLoggedErrForStatus map[retry.Status]struct{}
+)
+
+func init() {
+	hasLoggedErrForStatus = make(map[retry.Status]struct{})
+}
+
+// RemoteNetTracerUtil wraps interactions with a remote network tracer service
+type RemoteNetTracerUtil struct {
+	// Retrier used to setup network tracer
+	initRetry retry.Retrier
+
+	socketPath string
+	httpClient http.Client
+}
+
+// SetNetworkTracerSocketPath provides a unix socket path location to be used by the remote network tracer.
+// This needs to be called before GetRemoteNetworkTracerUtil.
+func SetNetworkTracerSocketPath(socketPath string) {
+	globalSocketPath = socketPath
+}
+
+// GetRemoteNetworkTracerUtil returns a ready to use RemoteNetTracerUtil. It is backed by a shared singleton.
+func GetRemoteNetworkTracerUtil() (*RemoteNetTracerUtil, error) {
+	if globalSocketPath == "" {
+		return nil, fmt.Errorf("remote tracer has no socket path defined")
+	}
+
+	if globalUtil == nil {
+		globalUtil = newNetworkTracer()
+		globalUtil.initRetry.SetupRetrier(&retry.Config{
+			Name:          "network-tracer-util",
+			AttemptMethod: globalUtil.init,
+			Strategy:      retry.RetryCount,
+			// 10 tries w/ 30s delays = 5m of trying before permafail
+			RetryCount: 10,
+			RetryDelay: 30 * time.Second,
+		})
+	}
+
+	if err := globalUtil.initRetry.TriggerRetry(); err != nil {
+		log.Debugf("network tracer init error: %s", err)
+		return nil, err
+	}
+
+	return globalUtil, nil
+}
+
+// GetConnections returns a set of active network connections, retrieved from the network tracer service
+func (r *RemoteNetTracerUtil) GetConnections() ([]tracer.ConnectionStats, error) {
+	// Otherwise, get it remotely (via unix socket), and parse from JSON
+	resp, err := r.httpClient.Get(connectionsURL)
+	if err != nil {
+		return nil, err
+	} else if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("conn request failed: socket %s, url: %s, status code: %d", r.socketPath, connectionsURL, resp.StatusCode)
```

Comment:
would having the response body be useful for the error as well or will that be logged elsewhere?

Reply:
I dont think so since we control the server and know there isn't any useful data in the body :)